pkg/util/mount/mount_linux.go

```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+	if err != nil {
+		return false, err
+	}
+
+	// process /proc/xxx/mountinfo in backward order and find the first mount
+	// point that is prefix of 'path' - that's the mount where path resides
+	var info *mountInfo
+	for i := len(infos) - 1; i >= 0; i-- {
+		if strings.HasPrefix(path, infos[i].mountPoint) {
+			info = &infos[i]
+			break
+		}
+	}
+	if info == nil {
+		return false, fmt.Errorf("cannot find mount point for %q", path)
+	}
+
+	// parse optional parameters
+	for _, opt := range info.optional {
+		if strings.HasPrefix(opt, "shared:") {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+type mountInfo struct {
+	root, mountPoint string
+	// list of "optional parameters", mount propagation is one of them
+	optional []string
+}
+
+// getMountInfo reads /proc/xxx/mountinfo and makes sure it did not change
+// between reads. This protects us from kernel adding/removing lines there
+// between individual read() syscalls.
+func getMountInfo(filename string) ([]mountInfo, error) {
+	// Read the file until we get the same content twice
+	oldInfo, err := parseMountInfo(filename)
+	if err != nil {
+		return []mountInfo{}, err
+	}
+	for {
+		newInfo, err := parseMountInfo(filename)
+		if err != nil {
+			return []mountInfo{}, err
+		}
+		if reflect.DeepEqual(oldInfo, newInfo) {
+			// Content is the same, finish
+			return oldInfo, nil
+		}
+		// Content is different, continue in the loop and try again
+		oldInfo = newInfo
+	}
+}
+
+// parseMountInfo parses /proc/xxx/mountinfo.
+func parseMountInfo(filename string) ([]mountInfo, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return []mountInfo{}, err
+	}
+	scanner := bufio.NewReader(file)
+	infos := []mountInfo{}
+
+	for {
+		line, err := scanner.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		fields := strings.Fields(line)
+		if len(fields) < 7 {
+			return nil, fmt.Errorf("wrong number of fields in (expected %d, got %d): %s", 8, len(fields), line)
+		}
+		info := mountInfo{
+			root:       fields[3],
+			mountPoint: fields[4],
+			optional:   []string{},
+		}
+		for i := 6; i < len(fields) && fields[i] != "-"; i++ {
+			info.optional = append(info.optional, fields[i])
+		}
+		infos = append(infos, info)
+	}
+	return infos, nil
+}
+
+// doMakeShared is common implementation of MakeShared on Linux. It checks if
+// path is shared and bind-mounts it as shared if needed. mountCmd and mountArgs
+// are expected to contain mount-like command, doMakeShared will add '--bind
+// <path> <path>' and '--make-shared <path>' to mountArgs.
+func doMakeShared(path string, mountInfoFilename string, mountCmd string, mountArgs []string) error {
+	shared, err := isShared(path, mountInfoFilename)
+	if err != nil {
+		return err
+	}
+	if shared {
+		glog.V(4).Infof("Directory %s is already on a shared mount", path)
+		return nil
+	}
+
+	glog.V(2).Infof("Bind-mounting %q with shared mount propagation", path)
+	// mount --bind /var/lib/kubelet /var/lib/kubelet
```

Comment:
huh? `syscall.Exec` will replace `/usr/bin/kulebet` in current process with `/bin/mount` and it won't ever return. We need `fork()`+`exec()` and that's exactly what `os.Exec` does.